File: 0801-0900/0841-Keys and Rooms/0841-Keys and Rooms.py
import collections
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = [False]*len(rooms)
        Q = collections.deque([0])
        while Q:
            index = Q.popleft()
            visited[index] = True
            for k in rooms[index]:
                if not visited[k]:
                    Q.append(k)
        return all(visited)

class Solution2:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = [False]*len(rooms)
        St = [0]
        while St:
            index = St.pop()
            visited[index] = True
            for k in rooms[index]:
                if not visited[k]:
                    St.append(k)
        return all(visited)

File: 0801-0900/0841-Keys and Rooms/0841-Keys and Rooms.cpp
class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        unordered_set<int> visited;
        queue<int> Q;
        Q.push(0);
        visited.insert(0);
        while (!Q.empty()) {
            int index = Q.front();
            Q.pop();
            for (int k : rooms[index]) {
                if (!visited.count(k)) {
                    Q.push(k);
                    visited.insert(k);
                }
            }
        }
        return visited.size() == rooms.size();
    }
};
